### ISSUE TITLE
WPF and multiple WebView2s: Restoring an inactive tab causes an exception

### ISSUE THREAD

**Issue by champnic** [MicrosoftEdge/WebView2Feedback#637]
Finally! I was able to reliably repro the problem using `edge:\\discards` to manually control the tabs.
@champnic can you check this?

Steps:
1. Run the code below
2. Select `DISCARD ME tab`
3. Select `Active tab`
4. Click `[Urgent Discard]` for the tab named `DISCARD ME` (third row)
5. Select `DISCARD ME tab`
6. Exception is thrown 🥳 

Program.cs
```cs
using System;
using System.Windows;
using System.Windows.Controls;
using Microsoft.Web.WebView2.Wpf;

namespace Repro0x8007139F
{
    static class Program
    {
        [STAThread]
        static void Main()
        {
            /********* UI SETUP *********/
            new Application();

            var window = new Window
            {
                Width = 1500,
                Height = 700,
            };

            var grid = new Grid();
            grid.RowDefinitions.Add(new RowDefinition { Height = new GridLength(1, GridUnitType.Star) });
            grid.RowDefinitions.Add(new RowDefinition { Height = new GridLength(1, GridUnitType.Star) });
            window.Content = grid;

            var tabControl = new TabControl();
            Grid.SetRow(tabControl, 0);
            grid.Children.Add(tabControl);

            var discardsWebView = new WebView2();
            Grid.SetRow(discardsWebView, 1);
            grid.Children.Add(discardsWebView);
            /********* UI SETUP *********/

            discardsWebView.EnsureCoreWebView2Async().ContinueWith(_ =>
            {
                Application.Current.Dispatcher.Invoke(() => discardsWebView.Source = new Uri("edge://discards"));
            });

            var activeWebView = new WebView2();
            activeWebView.EnsureCoreWebView2Async().ContinueWith(_ =>
            {
                Application.Current.Dispatcher.Invoke(() =>
                {
                    activeWebView.NavigateToString(@"
                        <html>
                            <head>
                                <title>ACTIVE TAB</title>
                            </head>
                            <body style='background:green'>ACTIVE TAB</body>
                        </html>
                    ");
                });
            });

            tabControl.Items.Add(new TabItem { Header = "Active tab", Content = activeWebView });
            tabControl.Items.Add(new TabItem { Header = "DISCARD ME tab", Content = null });
            tabControl.SelectionChanged += (sender, e) =>
            {
                var selectedTabItem = (TabItem)e.AddedItems[0];
                if (selectedTabItem.Content != null)
                    return;

                var discardMeWebView = new WebView2();
                discardMeWebView.EnsureCoreWebView2Async().ContinueWith(_ =>
                {
                    Application.Current.Dispatcher.Invoke(() =>
                    {
                        discardMeWebView.NavigateToString(@"
                            <html>
                                <head>
                                    <title>DISCARD ME</title>
                                </head>
                                <body style='background:red'>DISCARD ME</body>
                            </html>
                        ");
                    });
                });
                selectedTabItem.Content = discardMeWebView;
            };

            window.ShowDialog();
        }
    }
}
```

Repro.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <UseWPF>true</UseWPF>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Web.WebView2" Version="1.0.674-prerelease" />
  </ItemGroup>
</Project>
```

_Originally posted by @AbeniMatteo in https://github.com/MicrosoftEdge/WebView2Feedback/issues/289#issuecomment-728826186_

[AB#30673255](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/30673255)

**Comment by champnic** [comment]
Other relevant info from @Abenimatteo:
Pure WPF TabControl that handles multiple webviews (one per tab).
This exception is thrown after an inactive tabs is restored (by clicking on the relative TabItem) but i'm not able to create a concrete repro.

`1.0.674-prerelease` doesn't solve the problem for me and disposing the control is not an option.

I thought it was caused by some resize\fill logic fired by attaching\detaching the control (example below) but it seems time related (inactivity).
```cs
private void TabControl_SelectionChanged(object sender, SelectionChangedEventArgs e)
{
    var webView = ...;

    // detach the old webview
    container.Children.Clear();
    // attach the new view
    container.Children.Add(webView);
}
```

**Comment by champnic** [comment]
Thanks for this repro @AbeniMatteo! Looks like the urgent discard is causing the render process to exit, so that the next time you try to access it (when switching to that tab) it fails. You can hookup the CoreWebView2.ProcessFailed event to catch when the connected processes exit.

Besides this repro, you were seeing this when you weren't doing the urgent discard? How long were the tabs in the background/inactive before you hit this? How many tabs did you have open?

It's possible we'll need to do work here to make sure WebView2 doesn't try to pre-emptively cleanup/discard tabs unless asked to.

**Comment by AbeniMatteo** [comment]
Additional infos:
- Only one tab is attached to the UI tree (unlike the repro), the others are not (a TabControl is used to selected which view\tab is displayed)
- The exception is thrown while manually switching tab, it is not thrown randomly and it cannot be handled with a try catch.
- I think more than ~20 mins of tab inactivity (it was difficult to repro because of this)
- Usually it happens while there are 2-4 tab open
- I don't think it's triggered by memory pressure, i'm usually at 50% ram usage

Considerations:
- Chrome allows extensions to [disable autodiscard (autoDiscardable field)](https://developer.chrome.com/extensions/tabs#method-update)

**Comment by champnic** [comment]
Ya I think it's reasonable for us to also have an API to set autodiscard to false (or really just make that the default for WebView2). I've added that info to the bug. In the meantime, I would suggest catching the process closures using the ProcessFailed event and then dispose/reload that WebView2.

**Comment by AbeniMatteo** [comment]
Calling dispose on WebView2 after OnProcessFailed sometimes throws the same exception 0x8007139F.
At least, this time, it seems catchable.

```
at Microsoft.Web.WebView2.Core.Raw.ICoreWebView2.remove_ContentLoading(EventRegistrationToken token)
at Microsoft.Web.WebView2.Core.CoreWebView2.remove_ContentLoading(EventHandler`1 value)
at Microsoft.Web.WebView2.Wpf.WebView2.Dispose(Boolean disposing)
at System.Windows.Interop.HwndHost.Dispose()
at _ProcessFailed(Object sender, CoreWebView2ProcessFailedEventArgs e)
at Microsoft.Web.WebView2.Core.CoreWebView2.OnProcessFailed(CoreWebView2ProcessFailedEventArgs args)
at Microsoft.Web.WebView2.Core.CoreWebView2ProcessFailedEventHandler.Invoke(ICoreWebView2 source, ICoreWebView2ProcessFailedEventArgs args)
```

**Comment by galpin** [comment]
Is there any update on adding support for disabling auto-discard? We suffer from the same problem.

**Comment by AbeniMatteo** [comment]
FYI, [this](https://github.com/MicrosoftEdge/WebView2Feedback/blob/2f527d46ed31dc1928d64281b2b6dd6227b2140f/specs/Freeze.md) doesn't help. I tried it but it affects suspended tabs and suspended != discarded.

**Comment by champnic** [comment]
This has been fixed in the latest stable runtime (92.0.902.51+). Auto-discarding has been disabled in WebView2. Thanks!
